### PR TITLE
Add an EffectAction class

### PR DESCRIPTION
**Pull request by zvonicek**
This PR contains a general implementation of the `ActionWithPredicate` protocol that takes a closure and triggers it when the provided effect or precidate is accepted and the action runs.

We identified a common pattern in our codebase of repeating simple, single-purpose `Action` classes eg. to handle various navigation events. This class helped us to generalize them. 

Support for effect cases with associated values does not come out of the box because of the Swift limitations, but it can be done by providing a custom predicate. Example of such use case is in unit tests.

### CODE REVIEW

File: MobiusExtras/Source/EffectAction.swift [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import MobiusCore
+
+/// A general implementation of the `ActionWithPredicate` protocol that takes a closure and triggers it when
+/// the provided effect or precidate is accepted and the action runs.
```
Comment (pettermahlen) [inline MobiusExtras/Source/EffectAction.swift]:
```suggestion
/// the provided effect or predicate is accepted and the action runs.
```

File: MobiusExtras/Source/EffectAction.swift [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+public class EffectAction<Effect>: ActionWithPredicate where Effect: Equatable {
+    public typealias Predicate = (Effect) -> (Bool)
+
+    private let predicate: Predicate
+    private let action: () -> Void
+
+    /// Initializes an EffectAction with a predicate.
+    ///
+    /// - Parameters:
+    ///   - predicate: a predicate whose accepting inputs the EffectAction accepts.
```
Comment (pettermahlen) [inline MobiusExtras/Source/EffectAction.swift]:
```suggestion
    ///   - predicate: a predicate selecting which Effects the EffectAction accepts.
```

File: MobiusExtras/Source/EffectAction.swift [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    ///   - action: a closure that gets triggered on the action run when the effect is accepted.
```
Comment (pettermahlen) [inline MobiusExtras/Source/EffectAction.swift]:
```suggestion
    ///   - action: a closure that gets executed when the EffectAction runs.
```

File: MobiusExtras/Source/EffectAction.swift [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    public init(predicate: @escaping Predicate, action: @escaping () -> Void) {
+        self.predicate = predicate
+        self.action = action
+    }
+
+    /// Initializes an EffectAction with an accepted effect.
+    ///
+    /// - Parameters:
+    ///   - acceptedEffect: en effect that the EffectAction accepts.
+    ///   - action: a closure that gets triggered on the action run when the effect is accepted.
```
Comment (pettermahlen) [inline MobiusExtras/Source/EffectAction.swift]:
```suggestion
    ///   - action: a closure that gets executed when the EffectAction runs.
```

File: MobiusExtras/Test/EffectActionTests.swift [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            context("when initializing with accepted effect") {
+                beforeEach {
+                    effectAction = EffectAction(.accepted, action: actionClosure)
+                    actionCalled = false
+                }
+
+                context("when asking for effect acceptance on acceptable effect") {
+                    it("is accepted") {
+                        expect(effectAction.canAccept(.accepted)).to(beTrue())
+                        expect(actionCalled).to(beFalse())
```
Comment (pettermahlen) [inline MobiusExtras/Test/EffectActionTests.swift]: What is the reason for validating that the action hasn't been called? That seems to be outside of the scope of the test case. Not a big deal, of course, but just a bit of extra noise in the tests.

File: MobiusExtras/Test/EffectActionTests.swift [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+                    }
+                }
+
+                context("when asking for effect acceptance on non acceptable effect") {
+                    it("is not accepted") {
+                        expect(effectAction.canAccept(.rejected)).to(beFalse())
+                        expect(actionCalled).to(beFalse())
```
Comment (pettermahlen) [inline MobiusExtras/Test/EffectActionTests.swift]: As above.

File: MobiusExtras/Test/EffectActionTests.swift [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                            return false
+                        },
+                        action: actionClosure)
+                    actionCalled = false
+                }
+
+                context("when asking for effect acceptance on acceptable effect") {
+                    it("is accepted") {
+                        expect(effectAction.canAccept(.acceptedWith(associatedValue: false))).to(beTrue())
+                        expect(actionCalled).to(beFalse())
```
Comment (pettermahlen) [inline MobiusExtras/Test/EffectActionTests.swift]: as above

File: MobiusExtras/Test/EffectActionTests.swift [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+                    }
+                }
+
+                context("when asking for effect acceptance on non acceptable effect") {
+                    it("is not accepted") {
+                        expect(effectAction.canAccept(.rejected)).to(beFalse())
+                        expect(actionCalled).to(beFalse())
```
Comment (pettermahlen) [inline MobiusExtras/Test/EffectActionTests.swift]: As above.

File: MobiusExtras/Test/EffectActionTests.swift (hunk 5 above)
Comment (zvonicek) [reply MobiusExtras/Test/EffectActionTests.swift]: My original intention was to put it there as a counterweight to `beTrue()` check of it in `when running the action`, but I agree it's not really needed -- I'll remove it.

**Comment by jeppes** [comment]
We are creating a single API for effect handlers which we hope to migrate to. The other methods of creating effect handlers will be deprecated. Once that is in place, I don't think this will be necessary.
@zvonicek

**Comment by zvonicek** [comment]
The proposed Effect Handler simplification makes this unnecessary, so I'm closing the PR.
